rust/processor/src/worker.rs

```diff
@@ -0,0 +1,894 @@
+                processor_name.to_string(),
+                transaction_filter,
+                pb_channel_txn_chunk_size,
+            )
+            .await
+        });
+
+        // Create a gap detector task that will panic if there is a gap in the processing
+        let (gap_detector_sender, gap_detector_receiver) =
+            kanal::bounded_async::<ProcessingResult>(BUFFER_SIZE);
+        let (
+            processor,
+            gap_detection_batch_size,
+            default_gap_detector_sender,
+            parquet_gap_detector_sender,
+        ) = if self.is_parquet_processor.unwrap_or(false) {
+            let processor = build_processor(
+                &self.processor_config,
+                self.per_table_chunk_sizes.clone(),
+                self.deprecated_tables,
+                self.db_pool.clone(),
+                Some(gap_detector_sender.clone()),
+            );
+            let gap_detection_batch_size: u64 = self.parquet_gap_detection_batch_size;
+
+            (
+                processor,
+                gap_detection_batch_size,
+                None,
+                Some(gap_detector_sender),
+            )
+        } else {
+            let processor = build_processor(
+                &self.processor_config,
+                self.per_table_chunk_sizes.clone(),
+                self.deprecated_tables,
+                self.db_pool.clone(),
+                None,
+            );
+            let gap_detection_batch_size = self.gap_detection_batch_size;
+
+            (
+                processor,
+                gap_detection_batch_size,
+                Some(gap_detector_sender),
+                None,
+            )
+        };
+
+        tokio::spawn(async move {
+            create_gap_detector_status_tracker_loop(
+                gap_detector_receiver,
+                processor,
+                starting_version,
+                gap_detection_batch_size,
+            )
+            .await;
+        });
+
+        // This is the consumer side of the channel. These are the major states:
+        // 1. We're backfilling so we should expect many concurrent threads to process transactions
+        // 2. We're caught up so we should expect a single thread to process transactions
+        // 3. We have received either an empty batch or a batch with a gap. We should panic.
+        // 4. We have not received anything in X seconds, we should panic.
+        // 5. If it's the wrong chain, panic.
+
+        info!(
+            processor_name = processor_name,
+            service_type = PROCESSOR_SERVICE_TYPE,
+            stream_address = self.indexer_grpc_data_service_address.as_str(),
+            concurrent_tasks,
+            "[Parser] Spawning concurrent parallel processor tasks",
+        );
+
+        let mut processor_tasks = vec![fetcher_task];
+        for task_index in 0..concurrent_tasks {
+            let join_handle: JoinHandle<()> = self
+                .launch_processor_task(
```

Comment:
I see this function takes two Options, one for the default sender and one for the parquet sender. I see we always have exactly one Some and one Option. Feels like launch_processor_task could be made more safe by encoding this in the type system, e.g. by using an enum:

enum GapDetectorSender {
    Normal(AsyncSender<ProcessingResult>),
    Parquet(AsyncSender<ProcessingResult>),
}